File: t/VB/MedicalViewerDemo/UI/SetNudgeShrinkActionDialog.vb
' *************************************************************
' Copyright (c) 1991-2019 LEAD Technologies, Inc.              
' All Rights Reserved.                                         
' *************************************************************
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms

Imports Leadtools.MedicalViewer

Namespace MedicalViewerDemo
   Partial Public Class SetNudgeShrinkActionDialog : Inherits Form
      Private cell As MedicalViewerMultiCell = Nothing
      Private viewer As MedicalViewer
      Private _owner As MainForm
      Private Function GetSeparatedText(ByVal sourceString As String) As String
         Dim result As String = sourceString.ToString()
         Dim index As Integer = 1

         Do While index < result.Length
            If Char.IsUpper(result.Chars(index)) Then
               result = result.Insert(index, " ")
               index += 1
            End If

            index += 1
         Loop
         Return result
      End Function

      Public Sub New(ByVal owner As MainForm)
         InitializeComponent()

         cell = MainForm.DefaultCell
         viewer = owner.Viewer
         _owner = owner

         If cell.IsValidForAction(MedicalViewerActionType.NudgeTool, MedicalViewerMouseButtons.Wheel) Then
            _cmbNudgeMouseButton.Items.Insert(4, "Wheel")
         End If
         If cell.IsValidForAction(MedicalViewerActionType.ShrinkTool, MedicalViewerMouseButtons.Wheel) Then
            _cmbShrinkMouseButton.Items.Insert(4, "Wheel")
         End If

         If cell.IsValidForAction(MedicalViewerActionType.NudgeTool, MedicalViewerActionFlags.Selected) Then
            _cmbNudgeApplyTo.Items.Add("Selected Cells")
            _cmbNudgeApplyTo.Items.Add("All Cells")
         End If
         If cell.IsValidForAction(MedicalViewerActionType.ShrinkTool, MedicalViewerActionFlags.Selected) Then
            _cmbShrinkApplyTo.Items.Add("Selected Cells")
            _cmbShrinkApplyTo.Items.Add("All Cells")
         End If

         If cell.IsValidForAction(MedicalViewerActionType.NudgeTool, MedicalViewerActionFlags.Selected) Then
            _cmbNudgeApplyingMethod.Items.Add("On Release")
         End If
         If cell.IsValidForAction(MedicalViewerActionType.ShrinkTool, MedicalViewerActionFlags.Selected) Then
            _cmbNudgeMouseButton.Items.Add("On Release")
         End If

         Dim actionFlags As MedicalViewerActionFlags = cell.GetActionFlags(MedicalViewerActionType.NudgeTool)
         If (actionFlags Or MedicalViewerActionFlags.OnRelease) = actionFlags Then
            _cmbNudgeApplyingMethod.SelectedIndex = 1
         Else
            _cmbNudgeApplyingMethod.SelectedIndex = 0
         End If

         If (actionFlags And MedicalViewerActionFlags.Selected) = MedicalViewerActionFlags.Selected Then
            _cmbNudgeApplyTo.SelectedIndex = 1
         ElseIf (actionFlags And MedicalViewerActionFlags.AllCells) = MedicalViewerActionFlags.AllCells Then
            _cmbNudgeApplyTo.SelectedIndex = 2
         Else
            _cmbNudgeApplyTo.SelectedIndex = 0
         End If

         _cmbNudgeMouseButton.SelectedIndex = CInt(cell.GetActionButton(MedicalViewerActionType.NudgeTool))


         actionFlags = cell.GetActionFlags(MedicalViewerActionType.ShrinkTool)
         If (actionFlags Or MedicalViewerActionFlags.OnRelease) = actionFlags Then
            _cmbShrinkApplyingMethod.SelectedIndex = 1
         Else
            _cmbShrinkApplyingMethod.SelectedIndex = 0
         End If

         If (actionFlags And MedicalViewerActionFlags.Selected) = MedicalViewerActionFlags.Selected Then
            _cmbShrinkApplyTo.SelectedIndex = 1
         ElseIf (actionFlags And MedicalViewerActionFlags.AllCells) = MedicalViewerActionFlags.AllCells Then
            _cmbShrinkApplyTo.SelectedIndex = 2
         Else
            _cmbShrinkApplyTo.SelectedIndex = 0
         End If

         _cmbShrinkMouseButton.SelectedIndex = CInt(cell.GetActionButton(MedicalViewerActionType.ShrinkTool))
      End Sub

      Private Sub _btnOK_Click(ByVal sender As Object, ByVal e As EventArgs) Handles _btnOK.Click
         Me.Close()
         _btnApply_Click(sender, e)
      End Sub

      Private Sub _btnApply_Click(ByVal sender As Object, ByVal e As EventArgs) Handles _btnApply.Click
         Dim applyingOperation As MedicalViewerActionFlags = CType(0, MedicalViewerActionFlags)

         Select Case _cmbNudgeApplyTo.SelectedIndex
            Case 0
               applyingOperation = MedicalViewerActionFlags.Active
            Case 1
               applyingOperation = MedicalViewerActionFlags.Selected
            Case 2
               applyingOperation = MedicalViewerActionFlags.AllCells
         End Select

         If _cmbNudgeApplyTo.SelectedIndex > 0 Then
            If _cmbNudgeApplyingMethod.SelectedIndex = 1 Then
               applyingOperation = applyingOperation Xor MedicalViewerActionFlags.OnRelease
            Else
               applyingOperation = applyingOperation Xor MedicalViewerActionFlags.RealTime
            End If

         End If

         If CType(_cmbNudgeMouseButton.SelectedIndex, MedicalViewerMouseButtons) = MedicalViewerMouseButtons.Left Then
            _owner.CurrentAction = MedicalViewerActionType.NudgeTool
            _owner.LeftButtonAction = MedicalViewerActionType.NudgeTool
            _owner.CheckMoveMarkersAction(False)
            _owner.CheckSelectPoints(False)
         End If
         If CType(_cmbNudgeMouseButton.SelectedIndex, MedicalViewerMouseButtons) = MedicalViewerMouseButtons.Right Then
            _owner.RightButtonAction = MedicalViewerActionType.NudgeTool
         End If

         cell.SetAction(MedicalViewerActionType.NudgeTool, CType(_cmbNudgeMouseButton.SelectedIndex, MedicalViewerMouseButtons), applyingOperation)

         For Each viewerCell As MedicalViewerBaseCell In viewer.Cells
            viewerCell.SetAction(MedicalViewerActionType.NudgeTool, CType(_cmbNudgeMouseButton.SelectedIndex, MedicalViewerMouseButtons), applyingOperation)
         Next viewerCell




         applyingOperation = CType(0, MedicalViewerActionFlags)

         Select Case _cmbShrinkApplyTo.SelectedIndex
            Case 0
               applyingOperation = MedicalViewerActionFlags.Active
            Case 1
               applyingOperation = MedicalViewerActionFlags.Selected
            Case 2
               applyingOperation = MedicalViewerActionFlags.AllCells
         End Select

         If _cmbShrinkApplyTo.SelectedIndex > 0 Then
            If _cmbShrinkApplyingMethod.SelectedIndex = 1 Then
               applyingOperation = applyingOperation Xor MedicalViewerActionFlags.OnRelease
            Else
               applyingOperation = applyingOperation Xor MedicalViewerActionFlags.RealTime
            End If

         End If
         If CType(_cmbShrinkMouseButton.SelectedIndex, MedicalViewerMouseButtons) = MedicalViewerMouseButtons.Left Then
            _owner.CurrentAction = MedicalViewerActionType.ShrinkTool
            _owner.LeftButtonAction = MedicalViewerActionType.ShrinkTool
            _owner.CheckMoveMarkersAction(False)
            _owner.CheckSelectPoints(False)
         End If
         If CType(_cmbShrinkMouseButton.SelectedIndex, MedicalViewerMouseButtons) = MedicalViewerMouseButtons.Right Then
            _owner.RightButtonAction = MedicalViewerActionType.ShrinkTool
         End If

         cell.SetAction(MedicalViewerActionType.ShrinkTool, CType(_cmbShrinkMouseButton.SelectedIndex, MedicalViewerMouseButtons), applyingOperation)

         For Each viewerCell As MedicalViewerBaseCell In viewer.Cells
            viewerCell.SetAction(MedicalViewerActionType.ShrinkTool, CType(_cmbShrinkMouseButton.SelectedIndex, MedicalViewerMouseButtons), applyingOperation)
         Next viewerCell
      End Sub

      Private Sub _btnCancel_Click(ByVal sender As Object, ByVal e As EventArgs) Handles _btnCancel.Click
         Me.Close()
      End Sub

      Private Sub _cmbApplyTo_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles _cmbNudgeApplyTo.SelectedIndexChanged, _cmbShrinkApplyTo.SelectedIndexChanged
         Dim cmBox As ComboBox = CType(sender, ComboBox)

         If cmBox.Name = "_cmbNudgeMouseButton" Then
            If ((_cmbNudgeMouseButton.Text = "Wheel") OrElse (_cmbNudgeApplyTo.SelectedIndex = 0)) Then
               _cmbNudgeApplyingMethod.Enabled = False
            Else
               _cmbNudgeApplyingMethod.Enabled = True
            End If
         ElseIf cmBox.Name = "_cmbShrinkMouseButton" Then
            If ((_cmbShrinkMouseButton.Text = "Wheel") OrElse (_cmbShrinkApplyTo.SelectedIndex = 0)) Then
               _cmbShrinkApplyingMethod.Enabled = False
            Else
               _cmbShrinkApplyingMethod.Enabled = True
            End If
         End If
      End Sub

      Private Sub _cmbMouseButton_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles _cmbNudgeMouseButton.SelectedIndexChanged, _cmbShrinkMouseButton.SelectedIndexChanged
         Dim cmBox As ComboBox = CType(sender, ComboBox)

         If cmBox.Name = "_cmbNudgeMouseButton" Then
            If ((_cmbNudgeMouseButton.Text = "Wheel") OrElse (_cmbNudgeMouseButton.SelectedIndex = 0) OrElse (_cmbNudgeApplyTo.SelectedIndex = 0)) Then
               _cmbNudgeApplyingMethod.Enabled = False
            Else
               _cmbNudgeApplyingMethod.Enabled = True
            End If
            If (_cmbNudgeMouseButton.SelectedIndex = 0) Then
               _cmbNudgeApplyTo.Enabled = False
            Else
               _cmbNudgeApplyTo.Enabled = True
            End If

            If _cmbNudgeMouseButton.Text = _cmbShrinkMouseButton.Text Then
               _cmbShrinkMouseButton.SelectedIndex = 0
            End If
         ElseIf cmBox.Name = "_cmbShrinkMouseButton" Then
            If ((_cmbShrinkMouseButton.Text = "Wheel") OrElse (_cmbShrinkMouseButton.SelectedIndex = 0) OrElse (_cmbShrinkApplyTo.SelectedIndex = 0)) Then
               _cmbShrinkApplyingMethod.Enabled = False
            Else
               _cmbShrinkApplyingMethod.Enabled = True
            End If
            If (_cmbShrinkMouseButton.SelectedIndex = 0) Then
               _cmbShrinkApplyTo.Enabled = False
            Else
               _cmbShrinkApplyTo.Enabled = True
            End If

            If _cmbNudgeMouseButton.Text = _cmbShrinkMouseButton.Text Then
               _cmbNudgeMouseButton.SelectedIndex = 0
            End If
         End If
      End Sub
   End Class
End Namespace

File: t/CS/MedicalViewerDemo/UI/SetNudgeShrinkActionDialog.cs
// *************************************************************
// Copyright (c) 1991-2019 LEAD Technologies, Inc.              
// All Rights Reserved.                                         
// *************************************************************
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;

using Leadtools.MedicalViewer;

namespace MedicalViewerDemo
{
   public partial class SetNudgeShrinkActionDialog : Form
   {
      MedicalViewerMultiCell cell = null;
      MedicalViewer viewer;
      MainForm _owner;
      private string GetSeparatedText(string sourceString)
      {
         string result = sourceString.ToString();
         int    index = 1;

         while (index < result.Length)
         {
            if (Char.IsUpper(result[index]))
            {
               result = result.Insert(index, " ");
               ++index;
            }
            
            ++index;
         }
         return result;
      }

      public SetNudgeShrinkActionDialog(MainForm owner)
      {
         InitializeComponent();

         cell = MainForm.DefaultCell;
         viewer = owner.Viewer;
         _owner = owner;

         if (cell.IsValidForAction(MedicalViewerActionType.NudgeTool, MedicalViewerMouseButtons.Wheel))
            _cmbNudgeMouseButton.Items.Insert(4, "Wheel");
         if (cell.IsValidForAction(MedicalViewerActionType.ShrinkTool, MedicalViewerMouseButtons.Wheel))
            _cmbShrinkMouseButton.Items.Insert(4, "Wheel");

         if (cell.IsValidForAction(MedicalViewerActionType.NudgeTool, MedicalViewerActionFlags.Selected))
         {
            _cmbNudgeApplyTo.Items.Add("Selected Cells");
            _cmbNudgeApplyTo.Items.Add("All Cells");
         }
         if (cell.IsValidForAction(MedicalViewerActionType.ShrinkTool, MedicalViewerActionFlags.Selected))
         {
            _cmbShrinkApplyTo.Items.Add("Selected Cells");
            _cmbShrinkApplyTo.Items.Add("All Cells");
         }

         if (cell.IsValidForAction(MedicalViewerActionType.NudgeTool, MedicalViewerActionFlags.Selected))
            _cmbNudgeApplyingMethod.Items.Add("On Release");
         if (cell.IsValidForAction(MedicalViewerActionType.ShrinkTool, MedicalViewerActionFlags.Selected))
            _cmbNudgeMouseButton.Items.Add("On Release");

         MedicalViewerActionFlags actionFlags = cell.GetActionFlags(MedicalViewerActionType.NudgeTool);
         if ((actionFlags | MedicalViewerActionFlags.OnRelease) == actionFlags)
            _cmbNudgeApplyingMethod.SelectedIndex = 1;
         else
            _cmbNudgeApplyingMethod.SelectedIndex = 0;

         if ((actionFlags & MedicalViewerActionFlags.Selected) == MedicalViewerActionFlags.Selected)
            _cmbNudgeApplyTo.SelectedIndex = 1;
         else if ((actionFlags & MedicalViewerActionFlags.AllCells) == MedicalViewerActionFlags.AllCells)
            _cmbNudgeApplyTo.SelectedIndex = 2;
         else
            _cmbNudgeApplyTo.SelectedIndex = 0;

         _cmbNudgeMouseButton.SelectedIndex = (int)cell.GetActionButton(MedicalViewerActionType.NudgeTool);


         actionFlags = cell.GetActionFlags(MedicalViewerActionType.ShrinkTool);
         if ((actionFlags | MedicalViewerActionFlags.OnRelease) == actionFlags)
            _cmbShrinkApplyingMethod.SelectedIndex = 1;
         else
            _cmbShrinkApplyingMethod.SelectedIndex = 0;

         if ((actionFlags & MedicalViewerActionFlags.Selected) == MedicalViewerActionFlags.Selected)
            _cmbShrinkApplyTo.SelectedIndex = 1;
         else if ((actionFlags & MedicalViewerActionFlags.AllCells) == MedicalViewerActionFlags.AllCells)
            _cmbShrinkApplyTo.SelectedIndex = 2;
         else
            _cmbShrinkApplyTo.SelectedIndex = 0;

         _cmbShrinkMouseButton.SelectedIndex = (int)cell.GetActionButton(MedicalViewerActionType.ShrinkTool);
      }

      private void _btnOK_Click(object sender, EventArgs e)
      {
         this.Close();
         _btnApply_Click(sender, e);
      }

      private void _btnApply_Click(object sender, EventArgs e)
      {
         MedicalViewerActionFlags applyingOperation = (MedicalViewerActionFlags)0;

         switch (_cmbNudgeApplyTo.SelectedIndex)
         {
            case 0:
               applyingOperation = MedicalViewerActionFlags.Active;
               break;
            case 1:
               applyingOperation = MedicalViewerActionFlags.Selected;
               break;
            case 2:
               applyingOperation = MedicalViewerActionFlags.AllCells;
               break;
         }

         if (_cmbNudgeApplyTo.SelectedIndex > 0)
         {
            if (_cmbNudgeApplyingMethod.SelectedIndex == 1)
               applyingOperation ^= MedicalViewerActionFlags.OnRelease;
            else
               applyingOperation ^= MedicalViewerActionFlags.RealTime;

         }

         if ((MedicalViewerMouseButtons)_cmbNudgeMouseButton.SelectedIndex == MedicalViewerMouseButtons.Left)
         {
            _owner.CurrentAction = MedicalViewerActionType.NudgeTool;
            _owner.LeftButtonAction = MedicalViewerActionType.NudgeTool;
             _owner.CheckMoveMarkersAction(false);
             _owner.CheckSelectPoints(false);
         }
         if ((MedicalViewerMouseButtons)_cmbNudgeMouseButton.SelectedIndex == MedicalViewerMouseButtons.Right)
         {
            _owner.RightButtonAction = MedicalViewerActionType.NudgeTool;
         }

         cell.SetAction(MedicalViewerActionType.NudgeTool, (MedicalViewerMouseButtons)_cmbNudgeMouseButton.SelectedIndex, applyingOperation);

         foreach (MedicalViewerBaseCell viewerCell in viewer.Cells)
         {
            viewerCell.SetAction(MedicalViewerActionType.NudgeTool, (MedicalViewerMouseButtons)_cmbNudgeMouseButton.SelectedIndex, applyingOperation);
         }




         applyingOperation = (MedicalViewerActionFlags)0;

         switch (_cmbShrinkApplyTo.SelectedIndex)
         {
            case 0:
               applyingOperation = MedicalViewerActionFlags.Active;
               break;
            case 1:
               applyingOperation = MedicalViewerActionFlags.Selected;
               break;
            case 2:
               applyingOperation = MedicalViewerActionFlags.AllCells;
               break;
         }

         if (_cmbShrinkApplyTo.SelectedIndex > 0)
         {
            if (_cmbShrinkApplyingMethod.SelectedIndex == 1)
               applyingOperation ^= MedicalViewerActionFlags.OnRelease;
            else
               applyingOperation ^= MedicalViewerActionFlags.RealTime;

         }
         if ((MedicalViewerMouseButtons)_cmbShrinkMouseButton.SelectedIndex == MedicalViewerMouseButtons.Left)
         {
            _owner.CurrentAction = MedicalViewerActionType.ShrinkTool;
            _owner.LeftButtonAction = MedicalViewerActionType.ShrinkTool;
            _owner.CheckMoveMarkersAction(false);
            _owner.CheckSelectPoints(false);
         }
         if ((MedicalViewerMouseButtons)_cmbShrinkMouseButton.SelectedIndex == MedicalViewerMouseButtons.Right)
         {
            _owner.RightButtonAction = MedicalViewerActionType.ShrinkTool;
         }

         cell.SetAction(MedicalViewerActionType.ShrinkTool, (MedicalViewerMouseButtons)_cmbShrinkMouseButton.SelectedIndex, applyingOperation);

         foreach (MedicalViewerBaseCell viewerCell in viewer.Cells)
         {
            viewerCell.SetAction(MedicalViewerActionType.ShrinkTool, (MedicalViewerMouseButtons)_cmbShrinkMouseButton.SelectedIndex, applyingOperation);
         }
      }

      private void _btnCancel_Click(object sender, EventArgs e)
      {
         this.Close();
      }

      private void _cmbApplyTo_SelectedIndexChanged(object sender, EventArgs e)
      {
         ComboBox cmBox = (ComboBox)sender;

         if (cmBox.Name == "_cmbNudgeMouseButton")
         {
            _cmbNudgeApplyingMethod.Enabled = ((_cmbNudgeMouseButton.Text == "Wheel") || (_cmbNudgeApplyTo.SelectedIndex == 0)) ? false : true;
         }
         else if (cmBox.Name == "_cmbShrinkMouseButton")
         {
            _cmbShrinkApplyingMethod.Enabled = ((_cmbShrinkMouseButton.Text == "Wheel") || (_cmbShrinkApplyTo.SelectedIndex == 0)) ? false : true;
         }
      }

      private void _cmbMouseButton_SelectedIndexChanged(object sender, EventArgs e)
      {
         ComboBox cmBox = (ComboBox)sender;

         if (cmBox.Name == "_cmbNudgeMouseButton")
         {
            _cmbNudgeApplyingMethod.Enabled = ((_cmbNudgeMouseButton.Text == "Wheel") || (_cmbNudgeMouseButton.SelectedIndex == 0) || (_cmbNudgeApplyTo.SelectedIndex == 0)) ? false : true;
            _cmbNudgeApplyTo.Enabled = (_cmbNudgeMouseButton.SelectedIndex == 0) ? false : true;

            if (_cmbNudgeMouseButton.Text == _cmbShrinkMouseButton.Text)
            {
               _cmbShrinkMouseButton.SelectedIndex = 0;
            }
         }
         else if (cmBox.Name == "_cmbShrinkMouseButton")
         {
            _cmbShrinkApplyingMethod.Enabled = ((_cmbShrinkMouseButton.Text == "Wheel") || (_cmbShrinkMouseButton.SelectedIndex == 0) || (_cmbShrinkApplyTo.SelectedIndex == 0)) ? false : true;
            _cmbShrinkApplyTo.Enabled = (_cmbShrinkMouseButton.SelectedIndex == 0) ? false : true;

            if (_cmbNudgeMouseButton.Text == _cmbShrinkMouseButton.Text)
            {
               _cmbNudgeMouseButton.SelectedIndex = 0;
            }
         }
      }
   }
}
